### DISCUSSION TITLE
Build problems for ugs-core with NetBeans IDE 12.6 and Java: 17.0.1

### DISCUSSION THREAD

**Discussion by wambecqa** [winder/Universal-G-Code-Sender#1813] [Help] [answered]
I am trying to build the ugs platform, starting from the zip file downloaded: Universal-G-Code-Sender-master.zip
I use netbeans with following settings:
Product Version: Apache NetBeans IDE 12.6
Java: 17.0.1; Java HotSpot(TM) 64-Bit Server VM 17.0.1+12-LTS-39
Runtime: Java(TM) SE Runtime Environment 17.0.1+12-LTS-39
System: Windows 10 version 10.0 running on amd64; Cp1252; en_US (nb)
User directory: C:\Users\alber\AppData\Roaming\NetBeans\12.6
Cache directory: C:\Users\alber\AppData\Local\NetBeans\Cache\12.6

I tried to build ugs-parent without any modifications of the source, and got following errors:
Errors: 
  FixturesTest.testArcFixtures:46->runAllFixturesInPath:99->runFixture:133->checkOrInitializeFixture:153->checkFixture:166 » UncheckedIO
  FixturesTest.testArcFixturesCoarse:57->runAllFixturesInPath:99->runFixture:133->checkOrInitializeFixture:153->checkFixture:166 » UncheckedIO
  FixturesTest.testArcWithMeshLevelerFixtures:67->runAllFixturesInPath:99->runFixture:133->checkOrInitializeFixture:153->checkFixture:166 » UncheckedIO
  FixturesTest.testRunFromFixtures:85->runAllFixturesInPath:99->runFixture:133->checkOrInitializeFixture:153->checkFixture:166 » UncheckedIO
  GUIBackendTest.setWorkPositionWithAdditionExpression:554 » NullPointer Cannot ...
  GUIBackendTest.setWorkPositionWithDivisionExpression:596 » NullPointer Cannot ...
  GUIBackendTest.setWorkPositionWithDivisionExpressionShouldConvertToWorkPositionUnits:638 » NullPointer
  GUIBackendTest.setWorkPositionWithDivisionExpressionWithoutValue:624 » NullPointer
  GUIBackendTest.setWorkPositionWithDivisionExpressionhouldConvertHashToWorkPositionUnits:610 » NullPointer
  GUIBackendTest.setWorkPositionWithExpressionShouldSetNegativePosition:540 » NullPointer
  GUIBackendTest.setWorkPositionWithExpressionShouldSetPosition:526 » NullPointer
  GUIBackendTest.setWorkPositionWithMultiplicationExpression:568 » NullPointer C...
  GUIBackendTest.setWorkPositionWithMultiplicationExpressionWithoutValue:582 » NullPointer
  GUIBackendTest.setWorkPositionWithSubtractionExpression:653 » NullPointer Cann...
  GUIBackendTest.setWorkPositionWithValueExpressionShouldSetPosition:512 » NullPointer
  
  After searching I "concluded" that the first four errors were also related to the warning given for surefire:
  --- maven-surefire-plugin:3.0.0-M5:test (default-test) @ ugs-core ---
file.encoding cannot be set as system property, use <argLine>-Dfile.encoding=...</argLine> instead

The other errors are due to "JavaScript" engine that was not found for some reason, probably because this was no longer in my java-jdk?

So for solving the first four errors, I inserted in the ugs-parent pom.xml file the lines for the maven-surefire-plugin:
        <configuration>
            <argLine>-Dfile.encoding=${project.build.sourceEncoding}</argLine>
        </configuration>
This effectively solved the first four errors and the surefire warning, but magically, two new errors were detected:
Errors: 
com.willwinder.universalgcodesender.AbstractControllerTest  Time elapsed: 0.159 s  <<< ERROR!
java.lang.ExceptionInInitializerError
	at com.willwinder.universalgcodesender.AbstractControllerTest.<clinit>(AbstractControllerTest.java:64)
Caused by: java.lang.reflect.InaccessibleObjectException: Unable to make protected final java.lang.Class java.lang.ClassLoader.defineClass(java.lang.String,byte[],int,int,java.security.ProtectionDomain) throws java.lang.ClassFormatError accessible: module java.base does not "opens java.lang" to unnamed module @6b2fad11
	at com.willwinder.universalgcodesender.AbstractControllerTest.<clinit>(AbstractControllerTest.java:64)

com.willwinder.universalgcodesender.AbstractControllerTest  Time elapsed: 0.16 s  <<< ERROR!
java.lang.NoClassDefFoundError: Could not initialize class com.willwinder.universalgcodesender.AbstractControllerTest

  GUIBackendTest.setWorkPositionWithAdditionExpression:554 » NullPointer Cannot ...
  GUIBackendTest.setWorkPositionWithDivisionExpression:596 » NullPointer Cannot ...
  GUIBackendTest.setWorkPositionWithDivisionExpressionShouldConvertToWorkPositionUnits:638 » NullPointer
  GUIBackendTest.setWorkPositionWithDivisionExpressionWithoutValue:624 » NullPointer
  GUIBackendTest.setWorkPositionWithDivisionExpressionhouldConvertHashToWorkPositionUnits:610 » NullPointer
  GUIBackendTest.setWorkPositionWithExpressionShouldSetNegativePosition:540 » NullPointer
  GUIBackendTest.setWorkPositionWithExpressionShouldSetPosition:526 » NullPointer
  GUIBackendTest.setWorkPositionWithMultiplicationExpression:568 » NullPointer C...
  GUIBackendTest.setWorkPositionWithMultiplicationExpressionWithoutValue:582 » NullPointer
  GUIBackendTest.setWorkPositionWithSubtractionExpression:653 » NullPointer Cann...
  GUIBackendTest.setWorkPositionWithValueExpressionShouldSetPosition:512 » NullPointer

Two questions:
1. Any idea on the failure to initialize AbstractControllerTest ?
2. Any idea how to work around or solve the "JavaScript" engine problem?

Thanks in advance for your help.

**Comment by breiler** [comment, marked as answer]
You need Java 8 or 13 to build and run UGS. 

We are using a OpenGL library that only works on those JVM:s on certain platforms so we are stuck with older versions of Java for the time being. This causes moment 22 problems in other areas which you are experiencing. Short description below.

1. I recall having this problem when trying to build with more recent JVM:s and got problems with some of the test frameworks that uses some reflection stuff in the language that no longer is supported.
2. We are using a JavaScript feature which is built into Java. This feature was removed as of Java version 15 but can be added with a JAR. However this JAR isn't compatible with the current version of Netbeans we are using.

**Reply by wambecqa** [reply]
Using java13 indeed solves most of the problems.  Thanks for the answer.
Is it documented somewhere what your standard development and test environment is for the master build?  Maybe I can adapt to that?

**Comment by wambecqa** [comment]
I downloaded java 13.0.2 from Oracle, installed it, and configured it as the java engine for the project in Netbeans.
Then I tried to re-compile, but:
for solving the test-file format errors, I inserted in the ugs-parent pom.xml file the lines for the maven-surefire-plugin:
        <configuration>
            <argLine>-Dfile.encoding=${project.build.sourceEncoding}</argLine>
        </configuration>
Then I got one failed module for cloud storage, during the test(see errors below).  I suspect this was due again to some file format problem, and I disabled those tests.  Then all works fine!

ugs-platform-plugin-cloud-storage: FAILED

Tests run: 4, Failures: 2, Errors: 0, Skipped: 0, Time elapsed: 4.302 s <<< FAILURE! - in com.willwinder.ugs.nbp.S3FileSystemViewTest
com.willwinder.ugs.nbp.S3FileSystemViewTest.testGetFiles  Time elapsed: 1.355 s  <<< FAILURE!
java.lang.AssertionError: 

Expected size: 2 but was: 1 in:
[D:\Universal-G-Code-Sender-master\Universal-G-Code-Sender-master\ugs-platform\ugs-platform-plugin-cloud-storage\s3:\test-files]
	at com.willwinder.ugs.nbp.S3FileSystemViewTest.testGetFiles(S3FileSystemViewTest.java:145)

com.willwinder.ugs.nbp.S3FileSystemViewTest.testUploadDownloadFile  Time elapsed: 0.044 s  <<< FAILURE!
org.junit.internal.ArrayComparisonFailure: array lengths differed, expected.length=16 actual.length=0; arrays first differed at element [0]; expected:<67> but was:<end of array>
	at com.willwinder.ugs.nbp.S3FileSystemViewTest.testUploadDownloadFile(S3FileSystemViewTest.java:102)
Caused by: java.lang.AssertionError: expected:<67> but was:<end of array>
	at com.willwinder.ugs.nbp.S3FileSystemViewTest.testUploadDownloadFile(S3FileSystemViewTest.java:102)
